### ISSUE TITLE
Log version numbers with discovered engines

### ISSUE THREAD

**Issue by smoyer64** [junit-team/junit5#590]
There have been several instance where users ended up with mismatched versions of the JUnit 5 artifacts, such as #586.  Debugging these problems would be easier if the INFO-level log message also printed out the version number of the engine.  So this message:

```
INFO: Discovered TestEngines with IDs: [junit-jupiter]
```

Would become:

```
INFO: Discovered TestEngines with IDs: [junit-jupiter-5.0.0-SNAPSHOT]
```

Note that the engine name can remain a constant but the version could be pulled programmatically from the JAR's manifest file.

**Comment by marcphilipp** [comment]
Good idea! Do you want to give it a shot?

**Comment by smoyer64** [comment]
I've made a couple other promises to the junit-team and need to catch up on those ... if this is *not* an M4 issue, feel free to assign it to me.

I know that there are some gotchas when parsing META-INF (we've done it for most of our REST services), so maybe I can short circuit the work a bit by getting Penn State to donate the resource scanning/matching code.

**Comment by marcphilipp** [comment]
You're right, it's more a nice-to-have. I'll add it to the 5.0 Backlog for now and make it up-for-grabs.

**Comment by sbrannen** [comment]
Providing the version number is a good idea; however, we cannot simply concatenate it with the ID, separated by a hyphen like `junit-jupiter-5.0.0-SNAPSHOT`. The reason is that _that_ is actually _not_ the engine's ID, and a user should be able to copy-n-paste an ID from the log to explicitly _include_ or _exclude_ a registered engine based on its official ID.

In other words, we would need to include additional information parenthetically -- for example: `[junit-jupiter (version: 5.0.0-M3), junit-vintage (version: 4.12.0-M3)]`

**Comment by sbrannen** [comment]
Slated for M4 due to work in #592.

**Comment by smoyer64** [comment]
@sbrannen I like that format much better.

For Maven, Gradle and Ant+Ivy users it would be really nice to have the entire GAV.  Unfortunately most of my configuration problems are that I didn't provide any engines!

**Comment by sbrannen** [comment]
Work on this issue is currently _in progress_ in #592.

**Comment by sbrannen** [comment]
Since we use the artifact ID for the `Implementation-Title` in the manifest for all of our published JARs, we can easily include the artifact ID in the log output; however, we will not be able to access the group ID via the package info.

@sormuras, would you mind updating #592 to generate the output as follows?

`[junit-jupiter (artifact ID: junit-jupiter-engine, version: 5.0.0-M3), junit-vintage (artifact ID: junit-vintage-engine, version: 4.12.0-M3)]`

@smoyer64, if you feel like submitting code to parse the correct manifest file, we could then add a custom manifest header (e.g., `Maven-Coordinates`) that contains the Group ID, Artifact ID, and Version for each JAR -- which we would populate accordingly during the Gradle build. Otherwise, I think having the artifact ID and version as in #592 is sufficient for the time being.

**Comment by sormuras** [comment]
@sbrannen That single logging line is getting convoluted, a bit. How about splitting it into multiple lines? Like:
```
Discovered TestEngines with IDs: [junit-jupiter, junit-vintage]
  junit-jupiter -> version: 5.0.0-M3, artifact ID: junit-jupiter-engine, group ID: ...
  junit-vintage -> version: 4.12.0-M3, artifact ID: junit-vintage-engine, group ID: ...
  spek -> ...
  ...
```

**Comment by smoyer64** [comment]
@sormuras I like the idea of one line per discovered engine but I'd like to confirm that you meant that each line would be issued via a separate log statement.  (Allowing newlines in log messages is a security issue in many static code analysis systems).

**Comment by sormuras** [comment]
@smoyer64 Good point. Refactoring my String concatenation code... before committing next iteration.

**Comment by sormuras** [comment]
Mh, the log level discussion in #513 indicates, that even the single line is too much information, when printed to `System.err` by default. 

Here's my current rendering so far, running a single test from within IDEA 2016.3.1:
* log level is configured at single place, namely `logDiscoveredTestEngines(Level level, Iterable<TestEngine> testEngines)`
* new details, like artifact id, group id or others can be integrated easily
* rolling back to single line log is also possible
```
Dec 10, 2016 2:22:56 PM org.junit.platform.launcher.core.ServiceLoaderTestEngineRegistry logDiscoveredTestEngines
INFO: Discovered TestEngines with IDs: [junit-jupiter, junit-vintage]
Dec 10, 2016 2:22:56 PM org.junit.platform.launcher.core.ServiceLoaderTestEngineRegistry logDiscoveredTestEngines
INFO: Details of junit-jupiter: [version: DEVELOPMENT]
Dec 10, 2016 2:22:56 PM org.junit.platform.launcher.core.ServiceLoaderTestEngineRegistry logDiscoveredTestEngines
INFO: Details of junit-vintage: [version: 4.12.0-M3]
```

**Comment by marcphilipp** [comment]
I would go with a single line.

**Comment by sormuras** [comment]
Single line is back:
```
Dec 10, 2016 4:57:34 PM org.junit.platform.launcher.core.ServiceLoaderTestEngineRegistry logDiscoveredTestEngines
INFO: Discovered TestEngines with IDs: [junit-jupiter (version: DEVELOPMENT), junit-vintage (version: 4.12.0-M3, artifact ID: junit-vintage-engine)]
```

**Comment by sbrannen** [comment]
This has been resolved in 716efc8c439ecc93ea8cf05019130d59bedb2b9a.
